### PR TITLE
[cumsum][CUDA][64-bit indexing] Add 64-bit indexing path for `cumsum`

### PR DESCRIPTION
**Pull request by eqy**
For #143486

Interestingly enough changing the indexing type seems to degrade performance when a larger width is not needed, even on small sizes, so making this a template param rather than forcing all cases to 64-bit

cc @ptrblck @msaroufim

### CODE REVIEW

File: aten/src/ATen/native/cuda/ScanUtils.cuh [hunk 1]
```diff
@@ -343,12 +343,12 @@ __device__ void tensor_kernel_scan_innermost_dim_impl(T* row_buf, T *tgt_, const
...
       // https://research.nvidia.com/publication/single-pass-parallel-prefix-scan-decoupled-look-back
-      for (uint32_t m = 0; m <= log_num_threads_x; ++m) {
+      for (index_t m = 0; m <= log_num_threads_x; ++m) {
         if (row_exists) {
-          uint32_t s = 1 << m; // s = 2 ^ m
-          uint32_t a = ((threadIdx.x >> m) << (m + 1)) | s; // a = (threadIdx.x / s) * (2 * s) + s
-          uint32_t ti = a + (threadIdx.x % s);
-          uint32_t si = a - 1;
+          index_t s = 1 << m; // s = 2 ^ m
+          index_t a = (index_t) ((threadIdx.x >> m) << (m + 1)) | s; // a = (threadIdx.x / s) * (2 * s) + s
```
Comment (malfet) [inline aten/src/ATen/native/cuda/ScanUtils.cuh:349]: Nit
```suggestion
          auto a = static_cast<index_t>((threadIdx.x >> m) << (m + 1)) | s; // a = (threadIdx.x / s) * (2 * s) + s
```

File: aten/src/ATen/native/cuda/ScanUtils.cuh [hunk 2]
```diff
@@ -317,10 +317,10 @@ __device__ void tensor_kernel_scan_innermost_dim_impl(T* row_buf, T *tgt_, const
 
     // Perform scan on one block at a time, keeping track of the total value of
     // all blocks processed so far.
-    for (uint32_t block_col = 0; block_col < row_size; block_col += 2 * num_threads_x) {
+    for (index_t block_col = 0; block_col < row_size; block_col += 2 * num_threads_x) {
       // Load data into shared memory (two values per thread).
-      uint32_t col1 = block_col + threadIdx.x;
-      uint32_t col2 = block_col + num_threads_x + threadIdx.x;
+      index_t col1 = block_col + (index_t) threadIdx.x;
```
Comment (malfet) [inline aten/src/ATen/native/cuda/ScanUtils.cuh:322]: `block_col` is `index_t`, so cast is unnecessary, is it?

**Review by malfet — APPROVED** [review]

File: test/test_torch.py [hunk 3]
```diff
@@ -1778,6 +1778,20 @@ def test_deterministic_cumsum(self, device):
             res_cpu = input.cpu().cumsum(dim)
             self.assertEqual(res0, res_cpu, atol=1e-3, rtol=1e-2)
 
+    @onlyCUDA
+    @largeTensorTest('24GB', device='cuda')
+    @largeTensorTest('24GB', device='cpu')
+    def test_cumsum_64bit_indexing(self, device):
+        b = torch.ones((2*4096*8, 100000), dtype=torch.float, device='cuda')
```
Comment (malfet) [inline test/test_torch.py:1785]:
```suggestion
        b = torch.ones(2 * 4096 * 8, 100000, dtype=torch.float, device='cuda')
```

File: aten/src/ATen/native/cuda/ScanUtils.cuh [hunk 4]
```diff
@@ -343,12 +343,12 @@ __device__ void tensor_kernel_scan_innermost_dim_impl(T* row_buf, T *tgt_, const
 
       // Parallel reduction with Sklansky method. The diagram can be seen on this paper:
       // https://research.nvidia.com/publication/single-pass-parallel-prefix-scan-decoupled-look-back
-      for (uint32_t m = 0; m <= log_num_threads_x; ++m) {
+      for (index_t m = 0; m <= log_num_threads_x; ++m) {
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/ScanUtils.cuh:346]: I don't think you ever need wide type here?

File: aten/src/ATen/native/cuda/ScanUtils.cuh [hunk 5]
```diff
@@ -343,12 +343,12 @@ __device__ void tensor_kernel_scan_innermost_dim_impl(T* row_buf, T *tgt_, const
...
       // Parallel reduction with Sklansky method. The diagram can be seen on this paper:
       // https://research.nvidia.com/publication/single-pass-parallel-prefix-scan-decoupled-look-back
-      for (uint32_t m = 0; m <= log_num_threads_x; ++m) {
+      for (int m = 0; m <= log_num_threads_x; ++m) {
         if (row_exists) {
-          uint32_t s = 1 << m; // s = 2 ^ m
-          uint32_t a = ((threadIdx.x >> m) << (m + 1)) | s; // a = (threadIdx.x / s) * (2 * s) + s
-          uint32_t ti = a + (threadIdx.x % s);
-          uint32_t si = a - 1;
+          index_t s = 1 << m; // s = 2 ^ m
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/ScanUtils.cuh:348]: similarly here, all these indices are uint32_t, they index into shared memory array that can't be large

**Pull request closed without merging**